File: web/registry/src/registry/componentHost.ts
import { nativeSource } from "./vendorSource";
import type { Component } from "./types";

const swiftuiComponentHostSwift = nativeSource("registry/swiftui/component-host/NmpComponentHost.swift");
const swiftuiComponentHostFixtureSwift = nativeSource("registry/swiftui/component-host/NmpComponentHostConformance.swift");
const composeComponentHostKotlin = nativeSource("registry/compose/component-host/NmpComponentHostProvider.kt");
const composeComponentHostFixtureKotlin = nativeSource("registry/compose/component-host/NmpComponentHostConformance.kt");

export const componentHostComponents: Component[] = [
  {
    slug: "component-host",
    routeId: "component-host",
    version: "0.1.0",
    description: "App-root host/provider for NMP registry component profile and embed context.",
    platforms: {
      swiftui: {
        status: "stable",
        installId: "swiftui/component-host",
        version: "0.1.0",
        dependencies: ["user-avatar", "content-kind-registry"],
        longDescription:
          "`NmpComponentHost` and `.nmpComponentHost(...)` bind `NostrProfileHost`, `EmbedEnvelopeSource`, `EventRefResolverProtocol`, and `NostrKindRegistry` once at the SwiftUI app root. Components below still read the existing lower-level environment values and own only visible resolve/release lifecycle.",
        files: [
          { source: "swiftui/component-host/NmpComponentHost.swift", target: "Components/NmpComponentHost/NmpComponentHost.swift", role: "source", content: swiftuiComponentHostSwift },
          { source: "swiftui/component-host/NmpComponentHostConformance.swift", target: "Components/NmpComponentHost/Fixtures/NmpComponentHostConformance.swift", role: "fixture", content: swiftuiComponentHostFixtureSwift },
        ],
        screenshots: [],
        customization: [
          "Create the concrete profile host, embed source, event-ref resolver, and kind registry in your app shell, then pass those app-owned bridge objects to `.nmpComponentHost(...)`.",
          "`refs.profile` remains the profile source and `refs.event.envelopes` remains a derived render sidecar; the host wrapper does not own kernel handles or event parsing.",
          "Install with `--with fixture` to copy the in-memory conformance harness for app tests that must run without a live kernel.",
        ],
      },
      compose: {
        status: "stable",
        installId: "compose/component-host",
        version: "0.1.0",
        dependencies: ["user-avatar", "content-kind-registry"],
        longDescription:
          "`NmpComponentHostProvider` binds `LocalNostrProfileHost`, `LocalResolvedEventEmbeds`, `LocalEventRefResolver`, and `LocalNostrKindRegistry` once at the Compose app root. Components below still consume the existing locals and own only visible resolve/release lifecycle.",
        files: [
          { source: "compose/component-host/NmpComponentHostProvider.kt", target: "Components/NmpComponentHost/NmpComponentHostProvider.kt", role: "source", content: composeComponentHostKotlin },
          { source: "compose/component-host/NmpComponentHostConformance.kt", target: "Components/NmpComponentHost/Fixtures/NmpComponentHostConformance.kt", role: "fixture", content: composeComponentHostFixtureKotlin },
        ],
        screenshots: [],
        customization: [
          "Create the concrete profile host, event-ref resolver, resolved-envelope map, and kind registry in your app shell, then pass those app-owned bridge objects to `NmpComponentHostProvider(...)`.",
          "`LocalResolvedEventEmbeds` mirrors derived `refs.event.envelopes`, not authoritative event rows; Rust/NMP still owns protocol parsing and projection production.",
          "Install with `--with fixture` to copy the in-memory conformance harness for app tests that must run without a live kernel.",
        ],
      },
    },
  },
];

File: crates/nmp-cli/tests/component_host.rs
//! Install coverage for the app-level native component host entries.

mod helpers;

use helpers::{nmp, TempDir};
use std::fs;

#[test]
fn add_swiftui_component_host_installs_host_and_dependencies() {
    let tmp = TempDir::new("swiftui-component-host");

    let out = nmp(tmp.path(), &["add", "component", "swiftui/component-host"]);
    assert!(
        out.status.success(),
        "nmp add component swiftui/component-host failed: {}",
        String::from_utf8_lossy(&out.stderr)
    );

    assert!(tmp
        .path()
        .join("Components/NmpComponentHost/NmpComponentHost.swift")
        .exists());
    assert!(tmp
        .path()
        .join("Components/NostrUser/NostrProfileHost.swift")
        .exists());
    assert!(tmp
        .path()
        .join("Components/NostrContent/EmbedHostEnvironment.swift")
        .exists());
    assert!(tmp
        .path()
        .join("Components/NostrContent/NostrKindRegistry.swift")
        .exists());
    assert!(!tmp
        .path()
        .join("Components/NmpComponentHost/Fixtures/NmpComponentHostConformance.swift")
        .exists());

    let lock = fs::read_to_string(tmp.path().join("nmp.components.lock")).unwrap();
    assert!(lock.contains("id = \"swiftui/user-avatar\""));
    assert!(lock.contains("id = \"swiftui/content-kind-registry\""));
    assert!(lock.contains("id = \"swiftui/component-host\""));
    assert!(lock.contains("NmpComponentHost.swift"));
}

#[test]
fn add_compose_component_host_installs_provider_and_dependencies() {
    let tmp = TempDir::new("compose-component-host");

    let out = nmp(tmp.path(), &["add", "component", "compose/component-host"]);
    assert!(
        out.status.success(),
        "nmp add component compose/component-host failed: {}",
        String::from_utf8_lossy(&out.stderr)
    );

    assert!(tmp
        .path()
        .join("Components/NmpComponentHost/NmpComponentHostProvider.kt")
        .exists());
    assert!(tmp
        .path()
        .join("Components/NostrUser/NostrProfileHost.kt")
        .exists());
    assert!(tmp
        .path()
        .join("Components/NostrContent/EmbeddedEvent.kt")
        .exists());
    assert!(tmp
        .path()
        .join("Components/NostrContent/NostrKindRegistry.kt")
        .exists());
    assert!(!tmp
        .path()
        .join("Components/NmpComponentHost/Fixtures/NmpComponentHostConformance.kt")
        .exists());

    let lock = fs::read_to_string(tmp.path().join("nmp.components.lock")).unwrap();
    assert!(lock.contains("id = \"compose/user-avatar\""));
    assert!(lock.contains("id = \"compose/content-kind-registry\""));
    assert!(lock.contains("id = \"compose/component-host\""));
    assert!(lock.contains("NmpComponentHostProvider.kt"));
}

#[test]
fn add_component_host_with_fixture_installs_conformance_rows() {
    let swift = TempDir::new("swiftui-component-host-fixture");
    let swift_out = nmp(
        swift.path(),
        &[
            "add",
            "component",
            "swiftui/component-host",
            "--with",
            "fixture",
        ],
    );
    assert!(
        swift_out.status.success(),
        "nmp add component swiftui/component-host --with fixture failed: {}",
        String::from_utf8_lossy(&swift_out.stderr)
    );
    let swift_fixture = swift
        .path()
        .join("Components/NmpComponentHost/Fixtures/NmpComponentHostConformance.swift");
    let swift_body = fs::read_to_string(swift_fixture).unwrap();
    assert!(swift_body.contains("refs.profile"));
    assert!(swift_body.contains("refs.event"));
    assert!(swift_body.contains("refs.event.envelopes"));
    assert!(swift_body.contains("NmpComponentHostConformanceHarness"));

    let compose = TempDir::new("compose-component-host-fixture");
    let compose_out = nmp(
        compose.path(),
        &[
            "add",
            "component",
            "compose/component-host",
            "--with",
            "fixture",
        ],
    );
    assert!(
        compose_out.status.success(),
        "nmp add component compose/component-host --with fixture failed: {}",
        String::from_utf8_lossy(&compose_out.stderr)
    );
    let compose_fixture = compose
        .path()
        .join("Components/NmpComponentHost/Fixtures/NmpComponentHostConformance.kt");
    let compose_body = fs::read_to_string(compose_fixture).unwrap();
    assert!(compose_body.contains("refs.profile"));
    assert!(compose_body.contains("refs.event"));
    assert!(compose_body.contains("refs.event.envelopes"));
    assert!(compose_body.contains("NmpComponentHostConformanceHarness"));
}
